<a href="https://colab.research.google.com/github/sakuronohana/my_datascience/blob/master/2_DS_Bootcamp_Python_Basic_%26_Advanced.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# DS Bootcamp Teil 1a - Python Basic & Advanced #

## DS Bootcamp Verzeichnis ##

[**Teil 1 - Data Science - Prozesse & Grundlagen**](https://github.com/sakuronohana/my_datascience/blob/master/1_DS_Bootcamp_Prozesse_%26_Grundlagen.ipynb)

[**Teil 1a - Data Science - Python Basic & Advanced**](https://github.com/sakuronohana/my_datascience/blob/master/2_DS_Bootcamp_Python_Basic_%26_Advanced.ipynb)

[**Teil 1b - Data Science - R Basic & Advanced**](https://github.com/sakuronohana/my_datascience/blob/master/3_DS_Bootcamp_R_Basic_%26_Advanced.ipynb)

[**Teil 2 - Data Science - ML - Datenbeschaffung & Datenaufbereitung**](https://github.com/sakuronohana/my_datascience/blob/master/4_DS_Bootcamp_ML_Datenbeschaffung_Datenaufbereitung.ipynb)

[**Teil 3 - Data Science - ML - Modellwahl**](https://github.com/sakuronohana/my_datascience/blob/master/5_DS_Bootcamp_ML_Modellwahl.ipynb)

[**Teil 4 - Data Science - ML - Modell Training, Testing & Optimierung**](https://github.com/sakuronohana/my_datascience/blob/master/6_DS_Bootcamp_ML_Modell_Train_Test_Optimize.ipynb)

[**Teil 5 - Data Science - ML - Modell Produktivsetzung, Überwachung & Pflege**](https://github.com/sakuronohana/my_datascience/blob/master/7_DS_Bootcamp_ML_Modell_Deploy_Monitor_Maintain.ipynb)

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/f/f8/Python_logo_and_wordmark.svg/460px-Python_logo_and_wordmark.svg.png" alt="Drawing" width="400">

##About Python##
Python ([ˈpʰaɪθn̩], [ˈpʰaɪθɑn], auf Deutsch auch [ˈpʰyːtɔn]) ist eine universelle, üblicherweise interpretierte höhere Programmiersprache. Sie hat den Anspruch, einen gut lesbaren, knappen Programmierstil zu fördern. So werden beispielsweise Blöcke nicht durch geschweifte Klammern, sondern durch Einrückungen strukturiert. Wegen ihrer klaren und übersichtlichen Syntax gilt Python als einfach zu erlernen.

Python unterstützt mehrere Programmierparadigmen, z. B. die objektorientierte, die aspektorientierte und die funktionale Programmierung. Ferner bietet es eine dynamische Typisierung. Wie viele dynamische Sprachen wird Python oft als Skriptsprache genutzt. Die Sprache weist ein offenes, gemeinschaftsbasiertes Entwicklungsmodell auf, das durch die gemeinnützige Python Software Foundation gestützt wird, die de facto die Definition der Sprache in der Referenzumsetzung CPython pflegt.

(Quelle: Wikipedia)

## Python Basics##

###Zahlen

Python beinhaltet verschiedene "Arten" (en.: types) von Zahlen. Wir werden uns hauptsächlich mit Ganzzahlen (en.: integers) und Kommazahlen (en.: floats) beschäftigen. Später unternehmen wir einen kurzen Abstecher zu den komplexen Zahlen (en.: complex numbers). Integers sind einfach ganze Zahlen, positiv und negativ. Zum Beispiel: 2 und -2 sind Integers. Floats sind Kommazahlen und in Python besonders, da sie einen Dezimalpunkt oder einen Exponenten (e) beinhalten, um die Zahl zu definieren. Zum Beispiel sind 2.0 und -2.1 Zahlen mit Dezimalpunkt. 4E2 (4 mal 10 hoch 2) ist ebenso ein Beispiel für eine Float-Zahl in Python

http://nbviewer.jupyter.org/github/datamics/Originale-Python-Bootcamp/blob/master/1%20Python%20Grundlagen/1-Grundlagen%20der%20Python%20Objekte%20und%20Datenstrukturen/1-Zahlen.ipynb

In [0]:
# Addition
1 + 2

In [0]:
# Subtraktion
2 - 1

In [0]:
# Multiplikation
1 * 2

In [0]:
# Division - Wird im Python3 im Float Type ausgegeben.
4/2

In [0]:
# Potenzen 
2**3

In [0]:
# Wurzel
2**0.5

In [0]:
# Reihenfolge der Operationen wird automatisch durch Python sichergestellt. Bsp. Klammer, Punkt vor Strich
(2 + 10) * 10 + 3

In [0]:
# Variablen zuweisen. Namen dürfen nicht mit Zahl beginnen bsp. 4a, es darf keine Leerzeichen bsp a 4 (a_4) und 
# spezielle Sonderzeichen enthalten.
a=5
a+a

In [0]:
# Variablen lassen sich in verschiedensten Varianten einsetzt 
einkommen = 100
steuersatz = 0.15
steuer = einkommen * steuersatz
steuer

### Strings

Strings werden in Python dazu benutzt, um Textinformationen, wie bspw. Namen, festzuhalten.
Genau genommen sind Strings in Python eine Sequenz. Das bedeutet, dass Python jedes einzelne Element in der Sequenz (in diesem Fall ein String) individuell beachtet. Ein Beispiel ist der String "hallo", den Python als Sequenz der einzelnen Buchstaben in einer bestimmten Reihenfolge speichert. Das hat zur Folge, dass wir später in der Lage sein werden zu indexieren und so bestimmte Buchstaben (z.B. den ersten oder letzten) aus einem String auszulesen.

Diese Idee einer Sequenz ist wichtig in Python und sie wird uns begleiten.

Zu den Sequenzen in Python gehören:

    Strings
    Listen
    Tupel
    Binärdaten


In [0]:
'hallo, ich bin ein Wort'

In [0]:
# Wenn innerhalb eines Strings Hochkommas verwendet werden, dann muss mit Anführungszeichen gearbeitet werden.
"Hallo ich bin's"

In [0]:
print("Hallo Welt!")
print("Hallo Mond")

In [0]:
# Möchte man den String verschiedene Zeilen verteilen dann benötigt man ein \n (Auf Mac mit Umschalt+Alt+7)
print("Eine neue Zeile \n in eine neue Zeile")

In [0]:
# Um einen Tab einzufügen muss man \t benutzen.
print("Hier kommt eine Tab \t in der Zeile")

In [0]:
# Um die länge bzw. Anzahl Zeichen in eine String zu ermitteln verwenden wir len
len("Hallo Welt")

In [0]:
# Die Ausgabe von bestimmten Zeichen wird mit einer eckigen Klammer und der entsprechenden Nummer augegeben
s = "Hallo Welt!"
print(s[0]) # Hiermit wird die Positon 0 bzw. das erste Zeichen der Variable s augegeben

In [0]:
# Es lässt sich natürlich auch eine ganze Zeichenkette von einer bestimmten Positon aus, ausgeben.
print(s[1:]) # Nach 0
print(s[:3]) # von 0 bis 3
print(s[:]) # ganzer String
print(s[-3:]) # von hinten angefangen
print(s[::2]) # nur jeder 2. Buchstabe wird ausgegeben

In [0]:
# Um einen String zu erweitern kann man das + Zeichen verwenden. 
s + " Hi Mond" # Achtung dies wir nicht unter der Variabel s gespeichert. 
sx = s + " Hi Mond" # nun ist es in eine neuen Variabel gespeichert.
sx

In [0]:
# Auch Multiplikation in den Strings ist möglich bzw. der String wird um die Anzahl im Multiplikator erhöht.
s = "Z"
s*10

In [0]:
# Der String ist eine Objekt für welche eine bestimmte Methode mit zusätzlichen Parametern ausgeführt wird
# Objekt.Methode(Parameter)
s = 'Hallo Welt'
s.upper() # Wandelt alle Zeichen im String in Grossbuchstaben

In [0]:
# Eine der wichtigsten Methoden ist split
s.split() # Fügen wir nichts in die Klammern, dann trennt Split einfach jedes Wort. 

In [0]:
# Wir können aber in der Klammer auch definieren wo der Split stattfinden soll
s.split ('W') # Python splittet jetzt den String bei W

### Ausgabe Formatierung

In [0]:
# Hier wird gezeigt in welchen verschiedenen Arten bzw. Methoden die Print Anweisung formatiert werden kann.
# Empfohlen Methode um die Strings zu bearbeiten ist .format(). Es wird her aber auch andere vorgestellt. 

In [0]:
# Das ist eine normale Ausgabe eines Strings
print("Das ist ein String")

In [0]:
# Wollen wir eine bestimmte Variable in einen String einführen so machen wir das wie folgt. Dazu verwenden wir das %s
s = "String"
print('Setze einen String mit einem mod und s ein: %s' %(s)) # 

In [0]:
print('Kommazahlen: %1.0f' %(13.144)) # In diesem Beispiel generieren wir einen Platzhalter für Float ohne Komma

In [0]:
# Nun kommen wir zur Umwandlungsmethode diese können wir mit %s (String) %r (Repräsentativ)
print('Hier ist eine Nummer: %s. Hier ist ein String: %s' %(123.1, 'hi'))

# In diesem Beispiel können wir auch %r anstelle %s verwendet. Kommt auf das gleiche Resultat

In [0]:
# Wir können hiermit natürlich auch eine Tuple (unveränderbare Liste)
print ('Erstes: %s, zweite %s, drittes %r' %(('Hi',3.14,22)))

In [0]:
# So weit so gut. Wie bereits erwähnt ist es am Besten, wenn man generell die Format Methode nimmt
print('Das ist ein String mit einer {p}'.format(p='Einfügung'))

In [0]:
# Auch mit der Format-Methode lassen sich mehrer Objekte analog einer Liste ausgeben.
print('Erste Ausgabe {a}, Zweite Ausgabe {b}, Dritte Ausgabe {c}'.format(a=1, b=2, c='Drei'))

# Die Reihenfolge der Ausgabe spielt dabei keine Rolle

### Listen

Auch schon mit normalen Strings können Sequenzen bzw. Auflistungen gemacht werden diese sind aber im Normalfall nicht änderbar. Listen hingegen lassen sich auf verschiedenen Arten einsetzen und lassen sich auch jederzeit beliebig ändern und erweitern.

In [0]:
# In Listen lassen sich verschieden Arten von Typen (String, Float usw) unterbringen
liste = [1,2,3,'String', 2.4]

In [0]:
# Die Listen lassen sich auf verschiedene Arten ausgeben

len(liste) # Zeigt die Länge eines Liste bzw. die in der Liste erfassten Elemente 
liste[2] # Zeigt das dritte Element, beginnend mit 0, aus der Liste
liste[:2] # Gibt alles bis zum Element 3 aus. Achtung ab beginnt ab 0
liste[2:] # Gibt alles ab dem Element 3 aus.
liste[1:3] # Gibt alles zwischen Element 2 und 4 aus

[2, 3]

In [0]:
#Die Liste kann auch beliebig erweitert werden. Die hier beschriebene Erweiterung ist aber nur Temporär
liste + ['Neues Element']

[1, 2, 3, 'String', 2.4, 'Neues Element']

In [0]:
# Um die Liste permanent zu erweitern muss eine neue Liste angelegt werden.
liste = liste + ['Neues Element']
liste

In [0]:
# Einen gängige Methode einen Wert einer Liste hinzuzufügen ist die Methode .append
liste.append('Ein neueres Element')
liste

In [0]:
# Um ein Element wieder aus der Liste zu entfernen kann die Methode .pop verwendet werden
liste.pop(0) # Wir entfernen hier die 1
liste

In [0]:
# Um ein Element aus einer Liste zu enfernen und in eine andere Liste zu transferieren 
pop = liste.pop(1)
pop

In [0]:
# Eine weitere Ausgabemöglichkeit ist die Berechnung 
liste * 2 # Gibt die Liste zweimal aus
liste[1] + 2 # Addiert das erste Elemente mit 2 resp. 1 + 2
liste[2] * 2 # Multipliziert das zweite Element mit 2

In [0]:
# Um die Reihenfolgen einer Liste zu ändern kann wie folgt vorgegangen werden.
neue_liste = ['a','s', 'd', 'f']
neue_liste.reverse() # Die neue Liste wird jetzt umgekehrt erstellt
neue_liste.sort() # Die umgekehrte Liste kann jettzt noch alfapethisch sortiert werden
neue_liste

In [0]:
# Es können natürlich auch Listen in Listen erstellt werden. Eine sogennante Matrix
list_1 = [1,2,3]
list_2 = [4,5,6]
list_3 = [7,8,9]
matrix = [list_1, list_2, list_3] # Wir fügen nun die Listen zu einer Matrix zusammen
matrix # Hier geben wir nun die Matrix mit ihren 3 Zeilen aus

In [0]:
# Die erstellte Matrix kann man sich jetzt wie folgt vorstellen:
#   y y y
# x 1 2 3
# x 4 5 6
# x 7 8 9
# x repräsentiert die Zeile und y die Spalte

# Für den Zugriff muss zuerst die Zeile und dann das Element angegeben werden
matrix[0] #Gibt die komplette erste Zeile (0) aus
matrix[0][1] #Gibt das zweite Element der ersten Zeile aus

In [0]:
# Wir erstellen nun aus der Matrix eine erste Spalte mit Hilfe einer for-Schleife
spalte_1 = [zeile[0] for zeile in matrix]
spalte_1

In [0]:
spalte_1[:2]

Das was Listen jedoch am interessanten macht, ist dass man in diesem Konstrukt alle Arten von Daten erfassen kann. So könnte man Beispielsweise eine Anzahl verschiedener Dataframes in eine Liste einfügen und entsprechenden abrufen.

In [0]:
#Laden wir zuerst mal ein paar Datensätze in verschieden Dataframes
import pandas as pd
import numpy as np

datloc1 = 'https://raw.githubusercontent.com/sakuronohana/my_datascience/master/datasets/car_data.csv'
datloc2 = 'https://raw.githubusercontent.com/sakuronohana/my_datascience/master/datasets/google_stock_data.csv'
datloc3 = 'https://raw.githubusercontent.com/sakuronohana/my_datascience/master/datasets/plane_crash_data.csv'

df_car = pd.DataFrame(pd.read_csv(datloc1, sep=';'))
df_stock = pd.DataFrame(pd.read_csv(datloc2))
df_plane = pd.DataFrame(pd.read_csv(datloc3))


In [0]:
# Nun packen wir alle DataFrames in eine Liste

df_list = [df_car,df_stock,df_plane]

# Nun können wir individuell auf die Dataframes in der Liste zugreifen

# Zeigt erstes Dataframe in Liste. Nur zur Erinnerung bei Python fängt alles mit 0 an
df_list[0]

# Zeigt erstes Dataframe und dessen erst Spalte
df_list[0].iloc[:,0]

# Ist der Name des ersten Features des ersten Dataframes bekannt, dann kann auch wie folgt darauf zugegriffen werden.
df_list[0].mpg

# oder so
df_list[0][df_car['mpg'] == 21.0]
df_list[2][df_plane['date'].str.contains('September')]

date  ...                                            summary
0     September 17, 1908  ...  During a demonstration flight, a U.S. Army fly...
1     September 07, 1909  ...  Eugene Lefebvre was the first pilot to ever be...
4     September 09, 1913  ...  The airship flew into a thunderstorm and encou...
7     September 03, 1915  ...  Exploded and burned near Neuwerk Island,  when...
9     September 24, 1916  ...  Shot down by British aircraft crashing in flames.
...                  ...  ...                                                ...
5154  September 06, 2014  ...  The passenger plane crashed into the jungle wh...
5155  September 20, 2014  ...  While approaching Port Moresby Jacksons Intern...
5173  September 05, 2015  ...  The air ambulance went missing while on a medi...
5174  September 15, 2015  ...  While on the way to a remote fishing site, the...
5237  September 28, 2018  ...  The aircraft was approaching for a landing at ...

[483 rows x 13 columns]

### Dictionaries

Dictionaries auch bekannt als Mapping sind Sammlungen von Objekten welche über einen Key gespeichert werden. Ein wesentlicher Unterschied zu den Listen ist, dass hier nicht über einen Index [0] usw. auf die Objekte zugegriffen werden kann sondern nur mittels Key. Zudem wird bei Dicts mit geschweiften Klammern gearbeitet. (Mac = alt + 8{ und 9}

In [0]:
d = {'Key':'Wert','Key2':'Wert2'}
d['Key'] # Der Zugriff auf den Wert wir nun mittels Key gemacht.

In [0]:
# In einen Dict können die Keys im Gegensatz zu den Werten in einen Dict nicht mehr verändert werden
# In einem Dict können alle Arten von Element verstaut werden. Der Zugriff kann dann gemäss der 
# verschiedenen Methoden durchgeführt werden. 
dict = {'K_Int': 1234, 'K_Float':12.33, 'K_String':"String", 'K_List':['A', 'B', 'C']}
dict['K_Int']
dict['K_List'][0] # Beispiel für den Zugriff auf Liste
dict['K_String'].upper() # Beispiel für Zugriff auf String mit entsprechenden Methoden
dict['K_Int']-3+5 # Beispiel einer arithmetischen Operation

In [0]:
# Ein Dictionary kann auch durch neue Werte ergänzt werden
dict['Tier'] = 'Hund'
dict['Zahl'] = 12
dict

In [0]:
# Dictionaries lassen sich auch gut verschachteln bzw. ein Dict kann wie eine Liste auch in einem Dict erstellt werden
dict2 ={'Key': {'Unterkey': {'Unterunterkey': 'Wert'}}}
dict2['Key']['Unterkey']

In [0]:
# Auch in einen Dictionary bestehen Methoden
#dict2. # mit TAB-Taste zeigen alle verfügbaren Methoden
dict2.items

In [0]:
dict2.values()

### Tuple

Tubel sind ähnlich wie Listen nur können sie nicht mehr verändert bzw. ergänzt werden. Deshalb verwendet man Tubel 
vor allem dann wenn es Werte gibt die nicht mehr geändert werden müssen Bsp. Wochentage, usw. Tubel werden aber im Gegensatz zu Listen viel weniger eingesetzt.

In [0]:
tub = ['Montag', 'Dienstag', 'Mittwoch']
tub[0] # Der Zugriff auf eine Element im Tubel ist wie bei einer Liste
tub[0:2] # Gibt alle Elemente von 0-1 aus


In [0]:
# Auch hier bestehen Methoden wieder tub. + TAB. Es werden zwar diverse Methoden angezeigt die meisten gelten jedoch
# für die Listen. Für Tubels bestehen lediglich zwei Methoden
tub.index('Montag') # Zeigt den Index also 0 des Elements an
tub.count('Montag') # Zeigt die Anzahl der gleichen Elemente im Tubel an 

In [0]:
# Und hier noch der Beweis, dass am Tubel kein weiteres Element angehängt werde kann.
tub.append['Donnerstag']

### Dateien

Python nutzt Datei Objekte um mit externen Dateien auf eine Computer zu interagieren. Grundsätzlich kann Python mit
einer Vielzahl von Dateien arbeiten wobei für das eine oder andere Format noch zusätzlich Library Module installiert 
werden muss.

Bevor wir nun Anfang mit Dateien zu arbeiten, wollen wir zuerst herausfinden in welchem Working Directory 
wir arbeiten dazu geben wir den folgenden Befehl ein

In [0]:
pwd

In [0]:
# Nun öffnen wir eine Datei und lesen sie
file = open('test.txt')
file.read()

In [0]:
# Eine Besonderheit im Python ist, dass wenn die Datei einmal gelesen wurde, diese nicht erneut gelesen werden kann.
# Dies hat damit zu tun, das der Cursor bei der letzten Lese-Operation am Ende der Datei verblieben ist. Ein er
file.read() # Gibt jetzt nur '' aus
file.seek(0) # Mit diesem Befehl kann man nun den Curson zurücksetzen
file.read() # Und danach funktioniert es wieder wie gewohnt.

In [0]:
# Als nächtes öffnen wir die Datei und schreiben etwas neues hinein
file = open('test.txt','w')
file.write('Das ist eine neue Zeile')

In [0]:
# Um den neuen Inhalt erneut lesen zu können müssen wir die Datei erneut im Read-Modus lesen
file = open('test.txt','r')
file.read()

In [0]:
# Nun lesen wir mit einer Interation (Schleife for) die einzelnen Zeilen aus
for zeile in open('test.txt'):
    print(zeile)


###Sets und Booleans

Ein Set ist eine Sammlung von einzigartigen Elementen. Ein Set wird mit dem Befehl set()angelegt. 

In [0]:
s = set()
s.add(1) # mit add lassen sich neue Elemente ins Set einführen.
s.add(2)
s.remove(2) # mit remove lassen sich die Elemente auch wieder löschen
s.

In [0]:
# Man kann Beispielsweise auch eine Liste in ein Set überführen. Dabei werden 
# dann aber die redundanten Elemente ausgefiltert.
list = [1,1,2,2,3,4,5,6,6]
set(list)

Booleans sind Wahrheitswerte die entweder True (1) oder False (0) sind. Booleans werden oft benutzt um beispielsweise eine Entscheidung in einer Schleife zu forsieren.

In [0]:
a = ()
if a == True:
    print('Wahr')
else:
    print('Nicht Wahr')

### Vergleichsoperatoren

In [0]:
2 == 2 # Wenn Wert der Operatoren gleich sind ergibt es true
2 != 1 # Wenn Wert der Operatoren nicht gleich sind ergibt es true
2 > 1 # Wenn ein Wert der Operatoren grösser ist ergibt es true
1 < 2 # Wenn ein Wert der Operatoren kleiner ist ergibt es true
2 >= 2 # Wenn ein Wert der Operatoren grösser oder gleich ist ergibt es true
1 <= 1 # Wenn ein Wert der Operatoren kleiner oder gleich ist ergibt 

In [0]:
# Vergleichsoperatoren kann man auch verknüpfen
1 < 2 < 3 # Ergibt true
1 < 2 and 2 < 3 # Das vorherige Beispiel kann man auch so schreiben
1 == 2 or 2 < 3 # Auch mit or kann man Operatoren verknüpfen

### Anweisungen

**If, Elif und Else**

if, elif und else werden verwendet um Aktionen abhängig von bestimmten Voraussetzungen oder Ergebnissen durchzuführen.
Sprachlich sieht das wie folgt aus:

"Falls (if) die Ampel rot anzeigt, da bleibe stehen und prüfen dein Mobil falls (elif) die Ampel orange oder grün anzeigt kannst du über die Strasse gehen falls (else) nichts von beidem zutrifft (Ampel defekt) schaue links und rechts und gehen dann über die Strasse"

In [0]:
r = 'rot'
g = 'grün'

i = input ('Gibt etwas ein: ') # Mit Input kann man einen Wert eingeben.

if i == r:
    print('Bleibe stehen und prüfe dein Mobile')
elif i == g:
    print('Du kannst über die Strasse gehen')
else:
    print('Die Ampel ist möglicherweise kaputt. Gehe wenn es frei ist')
    
    
    

**for Schleife

Eine for Schleife wird auch Loop genannt. Sie wird solange durchlaufen, bis ein definierte Wert erreich wurde. Die kann eine Zahl sein oder das Ende einer Lese-Operation in einer Datei.

Sprachlich ausgedrückt sieht das dann so aus:

"Das Buch hat 10 Seiten für (for) jede Seite (Item) im (in) Buch (Objekt) lese jeden Buchstaben bis du auf Seite 10 angekommen bist."

Der Syntax sieht wie folgt aus:

for item in object

Das item ist frei wählbar.

In [0]:
buch = [1,2,3,4,5,6,7,8,9,10] # Eine Liste wird erstellt
buch_sum = 0

for seite in buch:
    buch_sum += seite # Das += (Selbstzuweisung) ersetzt folgende Zeile buch_sum = buch_sum + Seite
    print ('Nun lese ich Seite', buch_sum)

# In dem Beispiel oben wird bei jeder Interation der vorherige Wert zum
# aktuelle Wert addiert, beginnend bei 0. bzw. 
# 0 + 1 = 1
# 1 + 2 = 3
# 3 + 3 = 6
# 6 + 4 = 10 usw.

In [0]:
# Ein guter Trick um nun beispielweise nur die geraden Zahlen auszugeben ist 
# der Einsatz von Modulo. Modulo (%) teil eine Zahl durch die andere und gibt
# dann nur den Resten aus. Hier Modulo mit for Schleife:

for seite in buch:
    if seite % 2 == 0:
        print('Nun lese ich nur jeweils die graden Seiten dieses Buchs', seite)

In [0]:
# Das ganze funktioniert auch mit Strings
for zeichen in 'String':
    print(zeichen)

In [0]:
# In Bezug auf Tuples hat die Schleife eine besonderer Bedeutung. So können
# die Tuple Elemente auf verschieden Arten entpackt werden.
l = [(2,4),(6,8),(10,12)] # Hier erstellen wir Tuple innerhalb einer Liste

for tup in l:
    print(tup)

In [0]:
for (t1,t2) in l:
    print(t1,t2)

In [0]:
# Auch in Zusammenhang mit Listen kann man die for Schleife verwenden
dict = {'k1':1,'k2':2,'k3':3}
for item in dict:
    print (item)

In [0]:
dict.items() # So lassen sich überigen die in einem Dictionary definierten 
# Element bzw. Items ausgeben.

In [0]:
# Diese Methode kann in einer Schleife ebenfalls verwendet werden.
for k,v in dict.items(): # Wir definieren zwei Item Key und Value
    print(k,v)

**while Scheifen

Die while Schleife gehört zu den allgemeinsten Wegen eine Interation durchzuführen. Eine while Schleife wird solange ausgeführt, wie ein Wert True ist. 

Sprachliche ausgedrückt sieht das so aus:

"Wenn (while) dein Bankkonto noch im plus ist, kannst du Geld vom Automaten abheben (Statement), ansonsten (else) musst du wieder Geld auf dein Bankkonto überweisen."

Der Syntax sieht wie folgt aus:

while test:
    code statement
else
    finales code statement

In [0]:
i = 1 # Wir definieren eine Initialwert
while i < 10: # Wir fordern die Scheife auf solange der Wert i kleiner als 10 ist duchzulaufen
    print('X ist aktuell: ',i)
    i+=1 # Damit der Wert i steigt lassen wir ihn in jeder Rund eine 1 dazuaddieren.
else:
    print('Fertig')

In [0]:
# Mit dieser Anweisung können auch noch andere Funktionen mit gegeben werden 
# Break: hier können wir aus einem Loop ausbrechen
# Continue: Geht an den Anfang des nächsten umfasseden Loops
# Pass: Tut gar nichts

x = 0
while x < 10:
    print(x)
    x+=1
    if x==3:
        print('x==3')
        break
    else:
        print('Fortsetzen..')
        continue
    print('nach continue')



Wichtig bei der While Schleife ist, dass folgender Befehl zu einer Endlosschleife führen würde. Also darauf achten nie den folgendne Befehl zu verwenden:
while True:
    print('Ach du Scheisse!')

### Range()

Die Funktion Range erlaubt es uns eine Liste von Nummern auszugeben und dabei einen Start- und Endpukt zu denfinieren. Zudem können wir die Schrittgrösse ausgeben. In Python3 verbindet man diese Funktion mit einer Liste.

Der Syntax gestaltet sich wie folgt:

range(start,ende,schrittgrösse)

In [0]:
range(10)
x = list(range(0,10,2))
print(x)

# Der letzte Wert wird nie ausgegeben

In [0]:
# Hier noch eine weitere Variante wie man einen Start- und Endwert definieren kann
start = 0
end = 20
x = list(range(start,end))
print(x)

In [0]:
# Im nächsten Beispiel können auch noch definieren in welchen Schritten die Ausgabe 
# erfolgen soll.
x = list(range(start,end,2)) # Hier wird in zweierschritten ausgegeben.
print(x)

In [0]:
# In Zusammenhang mit diesen Funktionen (Range, List, Directory) stellt sich
# auch die Frage wie man eine grosse Anzahl von Daten im Memory behalten kann.
# die Lösung ist eine Generator mit der Funktion Range

for num in range(0,10,2): # Der Syntax von Range range(Startwert, Endwert, Schritte)
    print(num)

In [0]:
# Zahlen bis 50 durch 3 teilbar machen und nur diese Ausgeben welche durch 3 Teilbar sind
[x for x in range(1,50) if x%3 == 0] # Hier setzen wir Modulo ein

### Listen Abstraktion (List comprehension)

Die Listen-Abstraktion, eigentlich auch im Deutschen besser als "List Comprehension" bekannt, ist eine elegante Methode Mengen in Python zu definieren oder zu erzeugen. Die List-Comprehension kommt der mathematischen Notation von Mengen sehr nahe. In Mathematik definiert man die Quadratzahlen der natürlichen Zahlen beispielsweise als { x2 | x ∈ ℕ } oder die Menge der komplexen ganzen Zahlen { (x,y) | x ∈ ℤ ∧ y ∈ ℤ }.
So kann man Beispielsweise eine Formel definieren und jeweils die einzelnen Werte in der Liste berechnen und auf diese zugreifen.

In [0]:
# Hier eingehens nochmal ein Beispiel wie man eine Liste erstellt
l = []
for buchstabe in 'wort':
    l.append(buchstabe)
print(l)

In [0]:
# Ein List comprehension erstellt man wie folgt
list = [buchstabe for buchstabe in 'Wort']
print(list)

In [0]:
list = [x**2 for x in range(1,11)] # Beispiel mit Wurzel
print(list)

In [0]:
list = [x for x in range(11) if x % 2 == 0] # Beispiel mit Modulo
print(list)

In [0]:
celsius = [0,10,21.2,36.7]

fahrenheit = [((9/5*temp+32)) for temp in celsius]
print(fahrenheit)

In [0]:
list = [x**2 for x in range(11)]
for zahl in x:
    print(zahl) # Nun können wir eine if-Anweisung einbauen und einen bestimmten Wert abfragen
    if zahl == 4:
        print ('Dies ist ein bestimmter Wert')
    else:
        break

### Methoden

Methoden sind Funktionen die in Objekte eingebaut werden. Diese Methoden führen spezifische Aktionen mit dem Objekt durch können Parameter aufnehmen und Funktionen wiedergeben.

Der generelle Syntax einer Methode sieht wie folgt aus:

object.method(arg1,arg2,arg3,...)

In [0]:
# Hier ein einfaches Beispiel einer Methode
l = [1,2,3]
l.append(4) # mit Tab können die verfügbaren Methoden bsp. append angesehen werden
print(l)

In [0]:
# Ein Möglichkeit die Bescheibung einer Methode auszugeben ist mit help
w = "Das ist eine Wort"
w.count('e') # Count gibt die Anzahl der gesuchten Werte aus
help(l.count)

### Funktionen

Formell betrachtet ist eine Funktion ein nützliches Hilfsmittel, dass 
mehrerer Anweisungen gruppiert, damit sie mehr als einmal ausgeführt werden können.
Man kann Funktionen so definieren, dass man sie innerhalb des Codes immer wieder verwenden kann.

Eine Funktion beginnt mit einer def Anweisung

In [0]:
def name_der_funktion(parameter1, parameter2):
    '''Hier kommt ein Kommentar'''
    return
name_der_funktion # mit Shift + Tab kann man nun die Funktion inkl Kommentar ausgeben
help(name_der_funktion) # oder aus so 

In [0]:
# Um die Funktion auszuführen muss wir folgt vorgegangen werden

name_der_funktion(1,2) # Mit der Funktion über geben wir die geforderten Parameter

In [0]:
# In diesem Beispiel definieren wir eine Funktion mit eine Parameter
def sag_hallo(name): # Der Parameter ist zu vergleichen mit einem Platzhalter
        print('Hallo',name)

In [0]:
sag_hallo('David') # Hier über geben wir einen Name für den Parameter

In [0]:
def add_num(num1,num2): # Bei diesem Beispiel definieren wir eine Funktion mit 2 Parameter
    return num1+num2 # Nun definieren wir was in der Funktion gemacht werden soll

In [0]:
add_num(4,5) # Nun übergeben wir zwei Parameter und die Funktion führt sie aus

# In diesem Fall wird nun der Wert 4 und 5 zusammengerechnet

In [0]:
# Die Ausgabe einer Funktion können wir auch als Variable speichern
ergebnis = add_num(3,7)
print(ergebnis)

In [0]:
# All diese funktioniert natürlich auch mit strings und in direkter Ausgabe
print(add_num('Halllo ', 'David'))

In [0]:
# In einer Funktion lassen sich ganze Routinen mit Schleifen-Anweisungen ausgeben

def ist_prim(num):
    for n in range(2,num):
        if num % n == 0:
            print('keine Primzahl')
            break
    else:
        print('Primzahl')

In [0]:
ist_prim(10)

In [0]:
# Das oben gezeigte Beispiel geht auch einfacher

import math # Hier importieren wir eine mathematisches Python Package 

def ist_prim(num):
    if num % 2 == 0 and num > 2: #
        return False
    for i in range(3, int(math.sqrt(num))+1, 2):
        if num % i == 0:
            return False
    return True


In [0]:
ist_prim(11)

### Lambda Ausdrücke

Lambda Ausdrücke sind sehr nützliche Tools im Pyhton. Sie erlauben es uns sogenannte "anonyme" Funktionen zu erstellen. Vereinfacht gesagt heisst das, dass wir schnell eine Funktion erstellen können ohne sie per def zu definieren. Ein Lambda Ausruck ist meist auf einer Zeile geschrieben und ist für das coden einfacher Aufgaben gedacht. Bei grösseren Aufgaben verwenden wir def.

In [0]:
# Um ein Gefühl für Lambda zu erhalten werden wir nun eine def Funktion in einen Lambda Ausdruck umwandeln.

def square(num):
    return num**2

square(2)

In [0]:
# Wir können dies natürlich auch noch vereinfachen und in einer einzige Zeile schreiben.
def square(num): return num**2

square(4)

In [0]:
# Wir werden nun diese def Beispiel in eine Lambda Funktion umwandeln.

square = lambda num: num**2 # Die return Anweisung ist bereits in Lambda beinhaltet
square(5)

In [0]:
# Auch die Modulo bzw. die Suche nach einer teilbaren Zahl lässt sich mittels Lambda 
# um einiges vereinfachen.

even = lambda x: x%2==0 # Lässt sich eine Wert durch zwei dividieren?
even(9)

In [0]:
# Auch mit Strings lassen sich in Lambda operieren.

erstes = lambda s: s[0] # Gibt den Buchstabe im Index 0 aus
umk = lambda s: s[::-1] # Gibt den String in umgekehrter Reihenfolge aus
erstes('halt'), umk('Hallo')

In [0]:
# Hier noch ein Lambda Beispiel mit mehreren Parameter
add = lambda x,y: x+y
add(2,4)

In [0]:
# Hier ein kleines Programm zur Umrechung von Grad & Fahrenheit

quest = input('Was möchtest du berechnen 1 = Fahrenheit / 2 = Grad: ')

if quest == '1':
    grad = input('Gebe den Wert in Grad Celsius ein: ')
    g2f = lambda g: g*2+32
    print(g2f(int(grad)), 'Fahrenheit')
elif quest == '2':
    fahr = input('Gebe den Wert in Fahrenheit ein: ')
    f2g = lambda f: (f-32)/2
    print(f2g(int(fahr)), 'Grad Celsius')
else:
    print('Falscher Wert, Program beendet!')
    

### Verschachtelte Anweisung und Geltungsbereiche

Jetzt, da wir unsere eigenen Funktionen zu schreiben gelernt haben, ist es wichtig zu verstehen, wie Python mit Variablen Namen umgeht. Wenn wir einen Variable Namen erstellen wird dieser im sogenannten Namensraum (en.: name space) gespeichert. Variablen Namen haben außerdem einen Geltungsbereich, welcher bestimmt, welche anderen Teile des Codes diesen Namen kennen.

Der Geltungsbereich der Variablen ist über die sogenannte LEGB Regel definiert:

L: Lokal - Namen die in irgendeiner Form innerhalb einer Funktion (def oder lambda) zugewiesen werden und nicht global in dieser Funktion definiert sind.

E: Einschließende Funktionen - Namen im lokalen Geltungsbereich aller einschließenden Funktionen (def oder lambda) von innen nach außen.

G: Globale (Module) - Namen die auf der höchsten Ebene einer Datei zugewiesen werden oder innerhalb einer Funktion als global ausgewiesen werden.

B: Bereits (in Python) eingebaute - Namen die bereits in die vorinstallierten Module wie range, open, usw. eingebaut sind.

Lokale Variablen innerhalb einer Funktion gelten nur in Zusammenhang mit dieser Funktion und nicht ausserhalb.


In [0]:
# Beispiel - Lokal

calc = lambda y,x: x+y # x und y sind lokale Variablen
print(calc(10,5))

In [0]:
x = 50

def funk(x):
    print('x ist ',x) # Hier gilt die Globale Variable
    x = 2 # Ab hier gilt diese Variable, aber nur innerhalb dieser Funktion
    print ('Lokales x ist jetzt', x)
    
funk(x)
print('x ist noch',x)


In [0]:
# Beispiel - Einschliessende Funktionen

name = 'Das ist ein globaler Name'

def gruss():
    # Hier eine Einschliessende Funkton
    name = 'David'
    
    def hallo():
        name = 'Hans' # Innerhalb einer Funktion wird immer der "innerste" Wert genommen
        print('Hallo '+ name)
    
    hallo()

gruss()
print (name) # Wie wir unschwer erkennen können wird auch diese Variable ausgegeben.

In [0]:
# Eine bereits eingebaute Variable ist beispielsweise 

test = 'Test'
print(len(test))

### Objektorientierte Programmierung (OOP)

In [0]:
# Fast alles in Python ist ein Objekt so auch Listen
l = [1,2,3]
l.count(2)
print(type(l))

In [0]:
# wie ach Functionen

def quadrat(num):
    return num**2
print(type (quadrat))

### Class (Klassen)

Klassen sind von Nutzer definierte Objekte. Klassen sind wie Blaupausen welche die Eigenschaften zukünfitger Objekte festhält. Instanzen sind spezifische Objekte werlche aus bestimmten Klassen geschaffen werden. Wie Beispielweise die Liste welche wir oben erstellt haben.

In [0]:
# Wir generieren nun eine eigene KLasse

class Beispiel (object): # Wichtig ist das der Klassenname mit einem Grossbuchstaben beginnt
    pass

x = Beispiel() # Hier verküpfen wir unsere Klasse mit einer Variabel

print(type(x)) # Nun können wir uns vergewissern, dass x eine Klasse ist


Der Klasse können jetzt natürlich noch Attribute und Methoden 
hinzugefügt werden. 

Ein Attribute ist dabei eine Charakteristik eines Objekts. Wenn wir Beispielsweise eine Klasse "Hund" Haben können wir ihr Attribute wie Hunderasse, Alter, Stammbaum usw. hinzufügen.

Eine Methode ist eine Operation welche wir auf ein Objekt anwenden können. Beispielswiese bei der Klasse "Hund" würde dies bellen, knurren, springen usw. sein. 

In [0]:
# Hier das Beispiel Klasse "Hund"

class Hund(object):
    def __init__(self, rasse): # Hiermit definieren wir eine Funkton mit dem Attribut rasse
        self.rasse = rasse # Hiermit weisen wir die Funktion Rasse der Klasse zu
        
sam = Hund(rasse='Lab') # Mit Hilfe der Klasse generieren wir nun ein neues Objekt
frank = Hund(rasse='Husky') # Dito
frank # Wir können nun überprüfen zu welche Klasse das Objekt Frank gehört
print(type(frank)) # Das geht natürlich auch so


In [0]:
# Nun wollen wir das Objekt Sam mal mit einem in der Klasse definierten Attribut ausgehben
sam.rasse #Ein Klammer ist nicht notwendig, da es sich nicht um eine Methode handelt 

In [0]:
# Kommen wir nun zu den Klassen-Objekt-Attribute. Diese gelten für alle Instanzen einer Klasse
class Hund(object):
    spezies = 'Säugetier'
    
    def __init__(self,rasse,name):
        self.rasse = rasse
        self.name = name
        
sam = Hund('Lab','Sam')

print(type(sam))
sam.name
sam.spezies

### Methoden

Neben Attributen können auch Methoden (Funktionen) in einer Klasse definiert werden. Mit ihnen können Operationen mit den Attributen durchgeführt werden.


In [0]:
class Kreis(object):
    pi = 3.14
    
    def __init__(self, radius = 1): # Hier definieren wir wieder eine Funktion und weisen Radius gleich einen Default-Wert zu. 
        self.radius = radius
        
    def flaeche(self):
        return self.radius * self.radius * Kreis.pi
    
    def radiusEinstellen(self,radius): # Mit diesem Attribut können wir den Radius individuell verändern.
        self.radius = radius
        
    def radiusErhalten(self): #Mit diesem Attribut können wir die Einstellung jederzeit abfragen
        return self.radius
        

k = Kreis(radius=100)
k.flaeche() #Achtung da es sich um einen Methode handelt bnötigen wir wieder eine Klammer


In [0]:
k.radiusEinstellen(10)

In [0]:
k.flaeche()

In [0]:
k.radiusErhalten()

### Vererbung

Mit der Vererbung können neue Klassen durch bereits existierenden Klassen erstellt werden. Dies neu erstellten Klassen werden abgeleitete Klassen genannnt. Die Klassen von welche wir Vererben heissen Basisklassen. Mit der Vererbung lassen sich somit einfach bestehende Codes wieder verwenden.

In [0]:
class Tier(object):
    def __init__(self):
        print('Tier erschaffen')
        
    def werBinIch(self):
        print('Tier')
        
    def essen(self):
        print('Essen')
        
class Hund(Tier): # Wir erstellen eine neue Klassen und Vererben die Tier Klasse
    def __init__(self):
        Tier.__init__(self) #Hier beziehen wir uns auf die Klasse Tier
        print('Hund erstellen')
        
    def werBinIch(self):
        print('Hund')
        
    def bellen(self):
        print('Wuff!')
    

In [0]:
t = Tier()
t.werBinIch()
t.essen()

In [0]:
h = Hund()
h.werBinIch()
h.bellen()
h.essen() # Dieses Attribut hat die Klasse Hund von der Klasse Tier vererbt bekommen.

### Spezielle Methoden

In [0]:
# Wir kennen ja die bereits die Methoden welche in Python standardmässig implementiert sind.
l = [1,2,3]
print(l)
len(l)

In [0]:
# Solchen Methoden können wir aber auch selber definieren.

class Buch(object):
    def __init__(self, titel, autor, seiten):
        print('Ein Buch wird geschrieben!')
        self.titel = titel
        self. autor = autor
        self.seiten = seiten
    
    def __str__(self):
        return "Titel:%s, Autor: %s, Seiten %s"% (self.titel, self.autor, self.seiten)
        
    def __len__(self):
        return self.seiten
    
    def __del__(self):
        print('Ein Buch wurde gelöscht')
        
buch = Buch('Python rockt!', 'Patrik Di Lena', 1000)
print (buch)
        

In [0]:
del(buch) # Wenn wir jetzt die Methode len ausführen wird Python das nicht interpretieren können

In [0]:
class Ubuntu:
    def __init__(self, ramsize):
        self.ram = ramsize
        self.type = 'server'
        
Ubuntu = Ubuntu(2000)

In [0]:
print(Ubuntu)

### Errors und Ausnahmen Handhabung

Es gibt unterschiedliche Fehlerquellen bei der Nutzung von Python. Man unterscheidet dabei zwischen folgenden Fehlerarten:

Syntaxfehler - Hierbei handelt es sich zum Beispiel um einen vergessenen Einschub, Tipfehler usw.
Semantische Fehler - Diese Fehler sind die gefährlichsten von allen Fehlern, da sie die Logik des erstellen Algorithmus betreffen. Diese Fehler werden meist erst später erstellt.

In [0]:
# Das ist ein typischer SyntaxError auch End of Line Error (EOL) genannt

print(Hello) # Vergessen von String-Hochkommas bei Print-Funktion

In [0]:
# Syntaxfehler kann man mit try und except Anweisungen abfangen.



##Python für Data Science Bootcamp ##

Wenn es um Data Science mit Python geht, dann gibt es verschiedene Python-Libraries wie bspw. Pandas, Numpy, Scikit-Learn, MatPlotLib die es sich lohn genauer anzusehen. 


###Pandas###
Pandas (bis Version 0.25) ist eine Programmbibliothek für die Programmiersprache Python, die Hilfsmittel für die Verwaltung von Daten und deren Analyse anbietet. Insbesondere enthält sie Datenstrukturen und Operatoren für den Zugriff auf numerische Tabellen und Zeitreihen.

Weitere Beispiele können auf dem folgend Link gefunden werden  [Pandas](https://chrisalbon.com/#articles)

**Daten erstellen, importieren & exportieren, zusammenführen, durchsuchen**

Mit Pandas besteht eine Vielzahl von Möglichkeiten Daten zu importieren und zu exportieren. 

Nachfolgenden ein paar Beispiele: 

####DataFrames erstellen####

In [0]:
# Mittels Pandas lassen sich einfach sogenannte Dataframes erstellen.
# Dataframes sind nichts anderes als eine 2-dimensionale Datenstruktur
# in Form von Spalten (Columns) und Zeilen (Rows). 

# Zuerst muss die Library geladen geladen werden
import pandas as pd
import numpy as np

# Nun laden wir Daten von einer Source und bringen diese gleich in ein Dataframe
datloc = 'https://raw.githubusercontent.com/sakuronohana/my_datascience/master/datasets/new_strada2_data_2738408.csv'
datraw = pd.DataFrame(pd.read_csv(datloc))

# Wir können aber auch eine leeres Dataframe erstellen ...
datdf = pd.DataFrame()
datdf2 = pd.DataFrame()

# ... und fügen gleich mal ein paar Spalten zu
datdf['Anrede'] = ['Herr', 'Frau']
datdf['Vorname'] = ['Marc','Elisa']
datdf['Nachname'] = ['Berger', 'Braun']

# Wir können dem leeren Dataframe natürlich auch Spalten anderer Dataframes hinzufügen
datdf['New_BFS'] = datraw['BFS_NR']

datraw

####Daten importieren & exportieren####

In [0]:
''' Es lassen sich Daten direkt von GitHub laden. Wichtig dabei ist zu beachten, den 
URL des RAW Views zu nehmen

Im folgenden Beispiel werden Daten von einer kalifornischen Volkszählung direkt 
vom Internet laden. '''

import pandas as pd

'''Bei Daten welche direkt von GitHub geladen werden ist es wichtig das der 
Anfang von der URL mit raw.githubusercontent.com beginnt.'''

housing_data = pd.read_csv('https://raw.githubusercontent.com/ageron/handson-ml/master/datasets/housing/housing.csv')

housing_data.head()

In [0]:
# Daten lassen sich ganz einfach aus verschiedenen Dateien oder sogar
# Datenbanken laden. Die einfachste Form ist aus CSVs wie im folgenden
# Beispiel gezeigt.

import pandas as pd # Zuerst muss die Library geladen geladen werden

# from .csv
data = pd.read_csv("https://raw.githubusercontent.com/sakuronohana/my_datascience/master/planecrashinfo.csv") # nun lesen wir ein csv

# mit .head werden die in der Klammer definierten Zeilen angezeigt, 
# dass genügt um zu überprüfen ob er Import funktioniert.
data.head(10) 
# Mit der Option data.tail(10) werden die letzten 10 Zeilen angezeigt

In [0]:
# Ein weiteres Datenformat welches sich einfach importieren lässt ist JSON

import pandas as pd

#from .json

datloc = 'https://raw.githubusercontent.com/sakuronohana/my_datascience/master/datasets/Solarenergiepotenziale_Gemeinden_Daecher_und_Fassaden_2019.01.01.json'

datload = pd.read_json(datloc)

''' Normalerweise lädt pandas die Daten ohne weiters und importiert sie in eine Tabellenformat.
Es könnt aber sein das noch der Parameter orient notwendig ist.'''

datload = pd.read_json(datloc, orient = 'columns')

datload.head()

MunicipalityNumber  ...                                        Methodology
0                   1  ...  https://www.uvek-gis.admin.ch/BFE/redirect/sol...
1                   2  ...  https://www.uvek-gis.admin.ch/BFE/redirect/sol...
2                   3  ...  https://www.uvek-gis.admin.ch/BFE/redirect/sol...
3                   4  ...  https://www.uvek-gis.admin.ch/BFE/redirect/sol...
4                   5  ...  https://www.uvek-gis.admin.ch/BFE/redirect/sol...

[5 rows x 12 columns]

In [0]:
# Wie bereits erwähnt lasssen sich auch Daten auch aus Datenbanken extrahieren
import pandas as pd
import sqlite3 as sql

consql = sql.connect('chinook.db')
impsqldat = pd.read_sql('select * from playlists', consql)

# Mehr zu diesem Thema unter https://www.dataquest.io/blog/python-pandas-databases/

In [0]:
# Mit Pandas lassen sich auch Daten direkt aus dem Internet importieren
# Im folgenden Beispiel importieren wir historische Exxon Finanzdaten (XOM)
# von Yahoo.com mittels einer API.

import pandas as pd
import datetime
import pandas_datareader.data as web 


# Nun definieren wir den Datum-Range 1.Januar 2010 bis Heute
start = datetime.datetime(2010, 1 ,1)
end = datetime.datetime.now()

# Als nächsten importieren wir die Daten in ein Dataframe (df)
df_exxon = web.DataReader("XOM", "yahoo",start, end) 


print(df_exxon.head())


In [0]:
# Der Export von Daten gestaltet sich mit Pandas ebenfalls sehr einfach
# In den folgenden Beispiel werden die Daten einerseits in ein CSV und
# andererseits in ein XLS exportiert.

import pandas as pd # Zuerst muss die Library geladen geladen werden


# from .csv
data = pd.read_csv("https://raw.githubusercontent.com/sakuronohana/my_datascience/master/planecrashinfo.csv") # nun lesen wir ein csv


# Exportieren von ausgewählten Spalten in ein Excel
data[['date', 'location', 'operator']].to_excel('plaincrash_short.xls')

# Exportieren aller Daten in ein CSV
data.to_csv('plaintcrash_short.csv')


In [0]:
"""Im folgenden Beispiel importieren wir Daten aus einem in ein PDF konvertiertes Excel""" 

#! pip install tabula-py # Zuerst müssen wir das Modul installieren
from tabula import *

data_pdf = read_pdf("plaincrash_short.pdf")
data_pdf

####DataFrames durchsuchen####

Dataframes in Pandas lassen sich auf verschiedene Arten durchsuchen.

In [0]:
# Suchen wir einen bestimmten Wert in einer Spalten, dann können wir diese wie folgt eingeben.

import pandas as pd

url = 'https://raw.githubusercontent.com/sakuronohana/my_datascience/master/datasets/plane_crash_data.csv'
dimport = pd.read_csv(url, sep=',')

dimport[dimport['time'] == '17:18']



date  ...                                            summary
0     September 17, 1908  ...  During a demonstration flight, a U.S. Army fly...
4855   December 03, 2007  ...  Crashed into the sea while en route. The pilot...

[2 rows x 13 columns]

In [0]:
# Durchsuchen eines DataFrames mittels verschiedenen Werten in verschiedenen
# Spalten

import pandas as pd

url = 'https://raw.githubusercontent.com/sakuronohana/my_datascience/master/datasets/plane_crash_data.csv'
dimport = pd.read_csv(url, sep=',')

''' Mit dem folgenden Code suchen wir nach zwei verschienden Werten in zwei Spalten mittels AND d.h. 
es werden nur Zeilen angegeben, welche beide Werte besitzen'''
#print ('Suchresultat mit AND-Operator')
print(dimport[dimport['date'].str.contains('July') & dimport['operator'].str.contains('United')].head())

'''Mit diesem Code suchen wir entweder nach dem einen oder anderen Wert in der selben Spalte'''
#print('Sucheresultat mit OR-Operator')
#print(dimport[dimport['date'].str.contains('July') | dimport['operator'].str.contains('United')].head())
select_list = ['July', 'United']
print(dimport[dimport['date'].str.contains('|'.join(select_list))].head())

''' Mit dem folgenden Code suchen wir nach Werten welche grösser oder klein sind'''
print ('Suchresultat mit <> Operatoren')
print (dimport[dimport['time'] > '17:18'].head())

Eine andere gute Methode um Datensätze zu durchsuchen ist mittels iloc und loc

In [0]:
# Durchsuchen von Zeilen
dimport.iloc[:1] # Zeile 0

# Mit loc kann der Index neu definiert werden. Im folgenden Beispiel defnieren wir Date als den neuen Index
dimport = dimport.set_index(dimport['date'])



date  ...                                            summary
date                                    ...                                                   
September 17, 1908  September 17, 1908  ...  During a demonstration flight, a U.S. Army fly...
September 07, 1909  September 07, 1909  ...  Eugene Lefebvre was the first pilot to ever be...
July 12, 1912            July 12, 1912  ...  First U.S. dirigible Akron exploded just offsh...
August 06, 1913        August 06, 1913  ...  The first fatal airplane accident in Canada oc...
September 09, 1913  September 09, 1913  ...  The airship flew into a thunderstorm and encou...

[5 rows x 13 columns]

In [0]:
''' In dem folgenden Beispiel wird gezeigt wie sich bestimmte Merkmale bzw. Informationen über das Datenset herauslesen lassen.
Hier nochmal das Beispiel der kalifornischen Volkszählung.'''

import pandas as pd

housing_data = pd.read_csv('https://raw.githubusercontent.com/ageron/handson-ml/master/datasets/housing/housing.csv')

# Gibt die Spaltennamen, Anzahl Zeilen, Anzahl Werte ungleich Null und Datentyp aus
housing_data.info()

# Zählt Anzahl Datenwerte einer bestimmten Spalte
housing_data["ocean_proximity"].value_counts()

# Mit der folgenden Methode lassen sich die numerischen Merkmale zusammenfassen
housing_data.describe()



####DataFrames gruppieren & sortieren####

In [0]:
''' Die Daten in einem DataFrame lassen sich auf verschiedene Arten gruppieren.'''

import pandas as pd

datimploc = 'https://raw.githubusercontent.com/sakuronohana/my_datascience/master/datasets/plane_crash_data.csv'
datload = pd.read_csv(datimploc)

# Gruppieren wir den Datesatz nach Flugzeugtyp und lassen uns gleich auch die Anzahl Datensätze anzeigen.
datload.groupby('aircraft_type').count()

# Es lassen sich auch mehrer Merkmale gleichzeitig gruppieren
datload.groupby(['date','aircraft_type']).count()

# Die Ausgabe der Gruppierung ohne zusätzliches Attribut wie .count() funktioniert wie folgt.
test = datload.groupby(['date','aircraft_type'])
test.first()


In [0]:
''' Wie schon bei Excel oder einem anderen Tabellenkonstrukt ist es auch hier möglich 
die Daten nach bestimmten Präferenzen zu sortieren.'''

import pandas as pd

datimploc = 'https://raw.githubusercontent.com/sakuronohana/my_datascience/master/datasets/plane_crash_data.csv'
datload = pd.read_csv(datimploc)

# Nachfolgend sortieren wir die Daten alphabetisch nach dem Merkmal Operator 
datload.sort_values(by='operator', ascending=True)

# Nachfolgend sortieren wir mehrere Merkmale in alphabetischer Anordung
datload.sort_values(by=['date','operator'], ascending=True)

# In den oberen Beispielen hat das Dataframe die Sortierung nicht beibehalten, diese können wir mit inplace=true ändern
datload.sort_values(by='operator', ascending=True, inplace=True)
datload.head()


date  ...                                            summary
2459       May 25, 1971  ...  The cargo plane lost an engine while taking of...
5030     April 21, 2011  ...  The airl taxi crashed about one minute after t...
5194     April 29, 2016  ...  The helicopter crashed on the rocky shores of ...
1612    August 11, 1958  ...  Crashed on approach 10 minutes from the airpor...
5026  February 22, 2011  ...  For unknown reasons, the small plane crashed i...

[5 rows x 13 columns]

####DataFrames merge, join & concatenate####

Bevor wir uns nun dem Beispielen für eine Datenzusammenführung (merge) und -verbindung (join) widmen. Ist es wichtigt die verschiedenen Arten der Zusammenführung zu verstehen. Nachfolgend eine kleine Übersicht:

![alt text](https://shanelynnwebsite-mid9n9g1q9y8tt.netdna-ssl.com/wp-content/uploads/2017/03/join-types-merge-names.jpg)

***Inner Merge / Inner Join***

Hiermit werden nur die Werte von zwei Datensätzen zusammengeführt, für welche auf beiden Seiten Werte bestehen. Dabei wir mittels "on" die zu referenzierende Spalte angegeben bzw. das Set der Werte angegeben für welche es auch bei zweiten Datensatz Werte geben muss. Pandas Code:

      pandas.merge(Datensatz1, Datensatz2, on= 'Ref.Spalte')

Diese Art der Zusammenführung oder Verbindung ist bei Pandas (wie auch SQL) Standard d.h. wenn keine weiternen Optionen angegeben werden führt Pandas zwei Datensätze mit dieser Art zusammen.

***Outer Merge / Outer Join***

Mit dem Outer Join/Merge werden sämtlche Datensätze beider Seiten zusammengeführt, ungeachtet dessen ob auf beiden Seiten ein Wert besteht. Wo kein Wert besteht wird der Platzhalter "NaN" eingefügt. Pandas Code:

      pandas.merge(Datensatz1, Datensatz2, on= 'Ref.Spalte', how = 'outer')

***Left Merge / Left outer join***

Hierbei werden die Spalten und Werte des linken Datensatzes behalten und durch Daten des rechten Datensatzes ergänzt oder mit NaN aufgefüllt.

     pandas.merge(Datensatz_Links, Datensatz_Rechts, on='Ref.Spalte', how = 'left')

***Right Merge / Right outer join***

Hierbei werden die Spalten und Werte des rechten Datensatzes behalten und durch Daten des linken Datensatzes ergänzt oder mit NaN aufgefüllt.

     pandas.merge(Datensatz_Links, Datensatz_Rechts, on='Ref.Spalte', how = 'right')

***Hinweis***: Mit dem Argument **on=** definieren ich die Spalten, welche in beiden Datensätzen bestehen mittels welchen zusammengeführt wird. Sobald der Output etwas wie das ausgibt **_x** oder **_y**, kann davon ausgegangen werden, dass eine Feature sowohl im einen wie auch im anderen Datensatz besteht. 

Link: https://www.shanelynn.ie/merge-join-dataframes-python-pandas-index-1/

In [0]:
''' Haben wir Daten in zwei verschiedene Dataframes importiert, dann können diese Einfach miteinander verketten (Concat) werden.
Diese Verkettung von Daten brauen wir bspw. wenn wir mehrere Logfiles mit gleichen Merkmalen haben und diese in einen Datensatz
verschmelzen wollen '''

import pandas as pd
import numpy as np

# Zuerst erstellen wir zwei Dataframes mittels zwei 2D Arrays mit Hilfe der Random Funkton von Numpy

df1d2 = pd.DataFrame (np.random.rand(100,2), columns=['A','B'])
df2d2 = pd.DataFrame (np.random.rand(100,2), columns=['C','D'])
df3d2 = pd.DataFrame (np.random.rand(100,2), columns=['A','B'])
print(df1d2.shape)
# Nun führen wir die Spalten (Axis=1) dieser zwei Dataframes mit Hilfe der Concat Funktion zusammen
df4 = pd.concat([df1d2,df2d2], axis=1)

# Nachfolgend verketten wir die Zeilen (Axis=0) diese zwei Dataframes
df4 = pd.concat([df1d2,df3d2], axis=0)
print(df4.shape)

(100, 2)
(200, 2)


In [0]:
#Im folgenden Beispiel führen wir die Datensätze mit Hilfe von MERGE 
# zusammen

import numpy as np
import pandas as pd

left = pd.DataFrame({'Key': ['K0', 'K1', 'K2', 'K3','K4'], 
                    'A': ['A0', 'A1', 'A2', 'A3', 'A4'],
                    'B': ['B0', 'B1', 'B2', 'B3', 'B4']}) 
  
right = pd.DataFrame({'Key': ['K6', 'K1', 'K2', 'K3','K4'], 
                      'A': ['A5', 'A1', 'A2', 'A3', 'A4'],
                      'C': ['C0', 'C1', 'C2', 'C3', 'C4'], 
                      'D': ['D0', 'D1', 'D2', 'D3', 'D4']}) 
                        
# Merging the dataframes                       
dfm1 = pd.merge(left, right, how ='inner', on= ['Key','A'])
print ('Datensatz nach Inner merge: \n', dfm1.head())

dfm2 = pd.merge(left, right, how ='outer', on= ['Key','A']).fillna('unknown')
print ('Datensatz nach Outer merge: \n', dfm2.head())

dfm3 = pd.merge(left, right, how ='left', on=['Key', 'A']).fillna('unknown')
print ('Datensatz nach Left merge: \n', dfm3.head())

dfm4 = pd.merge(left, right, how ='left', on=['Key','A']).fillna('unknown')
print ('Datensatz nach Right merge: \n', dfm4.head())

Datensatz nach Inner merge: 
   Key   A   B   C   D
0  K1  A1  B1  C1  D1
1  K2  A2  B2  C2  D2
2  K3  A3  B3  C3  D3
3  K4  A4  B4  C4  D4
Datensatz nach Outer merge: 
   Key   A   B        C        D
0  K0  A0  B0  unknown  unknown
1  K1  A1  B1       C1       D1
2  K2  A2  B2       C2       D2
3  K3  A3  B3       C3       D3
4  K4  A4  B4       C4       D4
Datensatz nach Left merge: 
   Key   A   B        C        D
0  K0  A0  B0  unknown  unknown
1  K1  A1  B1       C1       D1
2  K2  A2  B2       C2       D2
3  K3  A3  B3       C3       D3
4  K4  A4  B4       C4       D4
Datensatz nach Right merge: 
   Key   A   B        C        D
0  K0  A0  B0  unknown  unknown
1  K1  A1  B1       C1       D1
2  K2  A2  B2       C2       D2
3  K3  A3  B3       C3       D3
4  K4  A4  B4       C4       D4


In [0]:
#Im folgenden Beispiel führen wir die Datensätze mit Hilfe von JOIN 
# zusammen

import numpy as np
import pandas as pd

left = pd.DataFrame({'A': ['A0', 'A1', 'A2', 'A3'], 
                    'B': ['B0', 'B1', 'B2', 'B3']}, 
                    index = ['K0', 'K1', 'K2', 'K3']) 
  
right = pd.DataFrame({'C': ['C0', 'C1', 'C2', 'C3'], 
                      'D': ['D0', 'D1', 'D2', 'D3']}, 
                      index = ['K0', 'K1', 'K2', 'K3'])
                        
# join the dataframes                       
dfm1 = left.join(right)
print ('Datensatz nach Inner merge: \n', dfm1.head())
                    
dfm2 = right.join(left)
print ('Datensatz nach Inner merge: \n', dfm2.head())


**Ändern, Ersetzen, Einfügen und Löschen**

Mit Pandas können einzelne oder mehrere Werte, Spalten und Zeilen verändern, ersetzen, hinzufügen oder gar löschen.

Siehe nachfolgend ein paar Beispiele:


In [0]:
# Mit Pandas lassen sich auch Daten in verschiedene Formate konvertieren

import pandas as pd # Zuerst muss die Library geladen geladen werden


# from .csv
data = pd.read_csv("https://raw.githubusercontent.com/sakuronohana/my_datascience/master/datasets/plane_crash_data.csv") # nun lesen wir ein csv

# Konverterien bestimmter Spalten in String
data['time'] = data['time'].astype('str')

# Konvertieren von bestimmten Spalten in Integer
# data['ground'] = data['ground'].fillna(-1).astype(int)

# Konvertieren von bestimmten Spalten in Numeric
# data['time'] = pd.to_numeric(data['time'])

data.head(5)



In [0]:
# Wir können aus ein durch Panda importiertes Datenset bestimmte Werte 
# auslesen.

import pandas as pd # Zuerst muss die Library geladen geladen werden


# from .csv
data = pd.read_csv("https://raw.githubusercontent.com/sakuronohana/my_datascience/master/datasets/plane_crash_data.csv") # nun lesen wir ein csv

# Nachfolgend lesen wir einen bestimmten Wert aus der Spalte operator
# aus der das Wort Military beinhaltet.

data_selc = data['operator'].where(data['operator'] != 'Military')


data_selc.head(10)

In [0]:
# Bei der Datenselektion mit Pandas lassen sich bestimmte Werte in Spalten
# löschen. Hierzu wird drop verwendet.

import pandas as pd # Zuerst muss die Library geladen geladen werden


# from .csv
data = pd.read_csv("https://raw.githubusercontent.com/sakuronohana/my_datascience/master/datasets/plane_crash_data.csv") # nun lesen wir ein csv


value_list = ['Private'] # Liste 
null_value_list = ['?']

# Mit .drop können unnötige Informationen wir folgt ausblendet werden
data = data.drop("time", axis=1) # Ganze Spalten
data1 = data.drop(data.index[2:4]) # Einzelen oder mehrere Zeilen
# :4 = bis und mit Zeile 3, 4: = Ab Zeile 4, 2:4 = Zeile 2 und 3 etc.
data2 = data.operator != 'Private' # Zeigt die Anzahl Zeilen mit True or Flase
data3 = data[data.operator.isin(value_list)] # Nur Zeilen mit Werten aus Liste
data4 = data[~data.operator.isin(null_value_list)] # Nur Zeilen mit Werten aus Liste

# mit .head werden die ersten 5 Zeilen angezeigt, dass genügt um zu 
# überprüfen ob er Import funktioniert.
data4.head()


In [0]:
# Das folgende Beispiel zeigt, wie man mit Pandas Werte in einem Datenset
# ändern kann.

import pandas as pd # Zuerst muss die Library geladen geladen werden


# from .csv
data = pd.read_csv("https://raw.githubusercontent.com/sakuronohana/my_datascience/master/datasets/plane_crash_data.csv") # nun lesen wir ein csv

# Nachfolgend ersetzen wir einen bestimmten Wert im Dataset durch einen
# anderen.

data_rep = data.replace('?', 'Unknown')

# Mit Replace lassen sich auch mehrer Werte gleichzeitig ersetzen.
data['route'] = data['route'].replace(['?','Test flight'],['0','Test Flight']).head(50)

# Natürlich können auch Spalten unbenennt werden
data.rename(columns={'route':'Route','date':'Datum'}).head()



Datum  ...                                            summary
0  September 17, 1908  ...  During a demonstration flight, a U.S. Army fly...
1  September 07, 1909  ...  Eugene Lefebvre was the first pilot to ever be...
2       July 12, 1912  ...  First U.S. dirigible Akron exploded just offsh...
3     August 06, 1913  ...  The first fatal airplane accident in Canada oc...
4  September 09, 1913  ...  The airship flew into a thunderstorm and encou...

[5 rows x 13 columns]

In [0]:
# Das folgende Beispiel zeigt, wie man mit Pandas eine neue Spalte erstellt.

import pandas as pd # Zuerst muss die Library geladen geladen werden


# from .csv
data = pd.read_csv("https://raw.githubusercontent.com/sakuronohana/my_datascience/master/datasets/plane_crash_data.csv") # nun lesen wir ein csv

# Nachfolgend erstellen wir eine neue Spalte und fügen Werte in die
# Datenfelder

data['casualties'] = 'n.a'

data.head(10)

In [0]:
# Das folgende Beispiel zeigt, wie man mit Pandas eine bestehende 
# Spalte löschen kann.

import pandas as pd # Zuerst muss die Library geladen geladen werden


# from .csv
data = pd.read_csv("https://raw.githubusercontent.com/sakuronohana/my_datascience/master/planecrashinfo.csv") # nun lesen wir ein csv

# Nun löschen wir eine Spalte



**Korrelation**

Im Rahmen der Vorbereitungen zur Umsetzung eines ML-Projektes werden die beschafften Daten erkundet und in diesem Zusammenhang visualisiert um eine Korrelation bzw. Beziehung zwischen den verschiedenen Merkmalen herauszufinden.

In [0]:
'''Um eine Korrelation zwischen den Daten optimal zu visualisieren
eignet sich die Scatter_matrix von Panda am besten.'''

import pandas as pd
from pandas.plotting import scatter_matrix

housing_data = pd.read_csv('https://raw.githubusercontent.com/ageron/handson-ml/master/datasets/housing/housing.csv')

'''Nachfolgend nehmen wir die vier Merkmale aus dem importierten Datensat
in welchen wir eine Korrelation suchen möchten'''
attributes = ["median_house_value", "median_income", "total_rooms",
             "housing_median_age"]
scatter_matrix(housing_data[attributes], figsize=(12,8))

#### Pandas (ab Verison 1.0) ###

Die Pandas Version 1.0 bringt noch ein paar neue Funktion mit sich. Die oben beschriebenen Funktionen sind in der neuen Version immer noch verfügbar.

In [0]:
# Installation der neuen Pandas Version

!pip3 install --upgrade pandas

In [0]:
# Check neue Version

import pandas as pd
pd.__version__

'1.0.1'

In [0]:
# Erstellung von Markdown-Tabellen aus Pandas Dataframes #

import pandas as pd

# Wichtig: Es sollte noch die Python Bibliothek tabulate installiert sein, damit
# dieser Code funktioniert. Auf Colab ist diese Library bereits vorinstalliert
# Falls nicht einfach "!pip3 install tabulate in" in einer Tabelle ausführen. 


daturl = 'https://raw.githubusercontent.com/sakuronohana/my_datascience/master/datasets/car_data.csv'
data = pd.read_csv(daturl, sep=';')

# Erstellen wir nun eine Tabelle mit dem Kriterium von 6 Zylindern
dattab =data[data['cyl'] == 6][['brand','mpg']]

# Nun exportieren wird die Tabelle in eine Markdown-Tabelle
print(dattab.to_markdown()) 

|    | brand          |   mpg |
|---:|:---------------|------:|
|  0 | Mazda RX4      |  21   |
|  1 | Mazda RX4 Wag  |  21   |
|  3 | Hornet 4 Drive |  21.4 |
|  5 | Valiant        |  18.1 |
|  9 | Merc 280       |  19.2 |
| 10 | Merc 280C      |  17.8 |
| 29 | Ferrari Dino   |  19.7 |


Nun einfach noch den Ouput von vorherigen Code markieren und in eine Text-Zelle kopieren. Als Result erhalten wir nun eine Markdown Tabelle: 

|    | brand          |   mpg |
|---:|:---------------|------:|
|  0 | Mazda RX4      |  21   |
|  1 | Mazda RX4 Wag  |  21   |
|  3 | Hornet 4 Drive |  21.4 |
|  5 | Valiant        |  18.1 |
|  9 | Merc 280       |  19.2 |
| 10 | Merc 280C      |  17.8 |
| 29 | Ferrari Dino   |  19.7 |

In [0]:
# Datentyp object zu string Konvertieren

import pandas as pd
daturl = 'https://raw.githubusercontent.com/sakuronohana/my_datascience/master/datasets/car_data.csv'
data = pd.read_csv(daturl, sep=';')

# nun können wir das Merkmal brand in einen String Datentype konvertieren.
print ('Datentyp vor der Konvertierung: ', data.brand.dtypes)

print ('Datentyp nach der Konvertierung: ', data.brand.convert_dtypes().dtypes)


Datentyp vor der Konvertierung:  object
Datentyp nach der Konvertierung:  string


###Numpy###
NumPy ist eine Programmbibliothek für die Programmiersprache Python, die eine einfache Handhabung von Vektoren, Matrizen oder generell großen mehrdimensionalen Arrays ermöglicht. Neben den Datenstrukturen bietet NumPy auch effizient implementierte Funktionen für numerische Berechnungen an.

Hier noch der Link zum Cheat Sheet Numpy (https://s3.amazonaws.com/assets.datacamp.com/blog_assets/Numpy_Python_Cheat_Sheet.pdf)

**Arrays in Numpy**

Die Hauptspezialität von Numpy ist der Umgang mit Arrays. Arrays sind Listen in welchen sich eine Vielzahl von Werten in einer einzigen Variablen zusammenfassen lassen.  

Hier noch einmal zu Erinnerung wie sich die Arrays gestalten:

![alt text](https://cdn-images-1.medium.com/max/2000/1*Ikn1J6siiiCSk4ivYUhdgw.png)

In [0]:
# Nachfolgend ein Beispiel wie einfach Arrays mit Numpy erstellt werden können

import numpy as np
import matplotlib.pyplot as plt


'''Erstellt einen Array mit 3 Zeilen und 5 Spalten aufgefüllt mit Nullen
"a0" oder Einsen "a1"'''
a0 = np.zeros([3,5])
a1 = np.ones([3,5])

'''Hier ein Beispiel eines 1D Array definiert als Integer. Es emfpiehlt sich 
zur Erstellung von Arrays np.array zu nutzen, da diese Variante speicheroptimierter
ist als der herkömmliche array Befehl von Python '''
ad1 = np.array([1,2,3,4,5,6,7], dtype = int)

'''Hier ein Beispiel eines 2D Array als Dtype String definiert'''
ad2 = np.array([(1,2,3,4),('a','b','c','d')], dtype = np.str) # 2D Array
print('Hier ein 2D Array:', ad2)

# Abfrage der Anzahl Zeilen und Spalten ein Arrays
print("Der Array hat folgende Anzahl Zeilen,Spalten:", np.shape(ad2))

# Abfrage der Dimension des Arrays
print('Der Array hat folgende Anzahl an Dimensionen:',np.ndim(ad2))

# Abfrage der Anzahl Datenwerte in einem Array
print('Der Array hat folgende Anzahl Datenwerte (Spalte * Zeilen):', np.size(ad2))

# Abfrage des Datentypes im Array
print('Der Array hat folgenden Datentype definiert:', ad2.dtype.name)

''' Ein Array kann auch mit dem Befehl np.arange erstellt werden. Die nachfolgende
Codezeile ist so zu verstehen: np.arange ([Start array bei dieser Zahl],
[Stop array bei dieser Zahl], [Erstellen den Zahlenrange in dieser Anzahl Schritte])'''
np.arange(0,100,5)


In [0]:
# Reshape eines Arrays
arr = np.array([1,2,3,4,5,6,7,8,9,10,11,12,13,14,15])

# Ein eindimensionaler Array ist ein Vector. 
print('Das ist ein Vector: ',arr)

''' Wir konventieren nun den Vector in eine Matrix mit dem Befehl np.reshape.
dabei ist die erste Zahl immer x (rows) und die zweite Zahl y (columns). 
Achtung: Ein Reshape funktioniert nur mit der richtigen Anzahl. Bei einem Reshape
auf eine 2x3 Matrix muss die Anzahl Werte durch 2 teilbar sein.'''
print ('\n Das ist eine 5 x 3 Matrix:\n', arr.reshape(5,3))
print('\n Das ist eine 3 x 5 Matrix: \n', arr.reshape(3,5))

'''Wir können eine Matrix bspw. auch mit Nullen füllen. Dazu gibt es einen eigenen 
Numpy Befehl.'''

print ('\nDas ist eine 5 x 5 Matrix: \n', np.zeros((5,5)))

In [0]:
''' In den folgenden Beispielen werden wir noch Numpy Operationen kennen-
lernen mit welchen man Arrays mutieren kann''' 

import numpy as np

a2d = np.array([(2,4,6),(20,40,60)])

'''Die Änderung der Struktur (Shape) eines Arrays kann nur in Zusammenhang mit  
einer Erweitung der möglich Datenelemente durchgeführt werden'''

np.size(a2d)
np.arange(9).reshape(3,3)



In [0]:
# Numpy eignet sich auch ausgezeichnet zur Erstellung von Zufallszahlen'''

import numpy as np
import pandas as pd


# Erstellung von 1D Array mit Random-Generator von Numpy
ard0 = np.random.rand(100,1)

# Erstellung 2D Array mit Random-Generator von Numpy
ard2 = np.random.rand(100,2)

''' Wir können auch gleich ein Dataframe aus dem 2D Array generieren, damit wir
wir die Daten in einer Tabelle darstell können'''
dfd2 = pd.DataFrame(np.random.rand(100,2), columns = ['Y','X'])

dfd2

In [0]:
# Wir können auf jedes Element des Arrays zugreifen.

import numpy as np

a2d = np.array([(1,2,3,4),(10,20,30,40)])
print(a2d)

# Zugriff auf einen Wert in einer bestimmten Zeile/Spalte
print ("Zugriff auf den Datenwert in der Zeile 0, Spalte 1 gibt den Wert:", a2d[0,1])

# Zugriff auf alle Werte in der Spalte 1
a2d[0:,0]


In [0]:
# Ein geniale Methode um bestimmte Wert aus einem Array zu löschen geht so :

import numpy as np

a2d = np.array([(1,2,3,4),(10,20,30,40)])
print(a2d)

a2d = np.setdiff1d(a2d, 2)
print (a2d) # Das 2 wurde gelöscht

**Mathematische Operationen mit Numpy**

Zu den Kernkompetenzen von Numpy gehören mathematischne Operationen mit Arrays.

In [0]:

import numpy as np
import pandas as pd

ard2 = np.random.rand(100,2)
ard2m = pd.DataFrame(ard2, columns=['X', 'Y'])

np.subtract(ard2m)

**Spezielle Numpy Funktionen**

Numpy hat natürlich unzählige wertvolle  Funktionen. Nachfolgend ein paar Beispiele:

In [0]:
# Mit Numpy lassen sich randomisierte Daten leicht erzeugen.

import pandas as pd
import numpy as np

# Datenframe erstellen
# Startwert für Ergebnisse setzen

np.random.seed(25)

''' Nachfolgenden wird ein 2D-Array mit 3 Spalten und jeweils 10 Werte
generierte'''

df = pd.DataFrame(np.random.rand(10, 3), columns = ['A', 'B', 'C'])

# Füllt die letzte Zeile mit Not a Number (NaN) Werte auf 
df.iloc[-1] = np.nan
df
# Addiert jeweils den Wert 1 und die Felder mit NaN mit zusätzlich 9
df.add(1,fill_value = 9)

###MatPlotLib, Seaborn & Co.###

Wenn es um die Visualisierung analysierter Daten geht, dann kommt man an MatPlotLib oder Seaborn nicht vorbei. Die Visualisierung lässt sich mit dieser Library sehr einfach bewerkstelligen.


In [0]:
# Im folgenden Plot wird ein Linien-Diagramm definiert

import  matplotlib.pyplot as plt

# Wir definieren eine Datenrange für die x und y
x = [1,2,3]
y = [5,7,4]

x2 = [1,2,3]
y2 = [10,14,12]

# Wir 
plt.plot(x,y, label='First Line') 
plt.plot(x2,y2, label='Second Line')
plt.xlabel('Plot Number')
plt.ylabel('Important Var')

plt.title('Interesting Graph\nCheck it out')
plt.legend()
plt.show()


In [0]:
# Es gibt eine einfache Möglichkeit alle Daten in einem Datensatz in sogenannten
# Histogramme darzustellen. Mit einem Histogramm lässt sich eine einfache Überblick
# über die Datenverteilung pro Spalte erstellen.

import pandas as pd
import matplotlib.pyplot as plt

housing_data = pd.read_csv('https://raw.githubusercontent.com/ageron/handson-ml/master/datasets/housing/housing.csv')

housing_data.hist(bins=50, figsize=(20,15))
plt.show()

In [0]:
# Ein weiteres Tool für die Visualiserung heist Ploty. Mit Ploty kann man auch gute Dashboards erstellen. https://plot.ly/python/

import pandas as pd
import plotly.express as px

car_data = pd.read_csv('https://raw.githubusercontent.com/sakuronohana/my_datascience/master/datasets/car_data.csv',sep=';')
car_data

fig = px.scatter(car_data, x='mpg', y='hp', size='mpg', color='brand')
fig.show()


In [0]:
import pandas as pd
import plotly.express as px
import datetime

# Daten aus der Coronavirus Statistik John Hopkins Universität
cv19_conf_url = 'https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_time_series/time_series_19-covid-Confirmed.csv'
cv19_death_url = 'https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_time_series/time_series_19-covid-Deaths.csv'
cv19_recy_url = 'https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_time_series/time_series_19-covid-Recovered.csv'

cv19confdat = pd.read_csv(cv19_conf_url)
cv19deathdat = pd.read_csv(cv19_death_url)
cv19recy = pd.read_csv(cv19_recy_url)

cv19plot = px.scatter_geo(cv19confdat, lat='Lat', lon='Long',color='Country/Region', hover_name='Country/Region')

cv19plot


###Scikit-Learn###

Für das Maschinelle Lernen existieren im Python spezielle Module in welchen bestimmte ML-Modelle bereits abgebildet sind und ein paar Datensets für das Training  hinterlegt sind.

####Sample Datasets im Scikit-Learn 

In [0]:
# Um ein Datenset mit Hilfe von Scikit zu laden muss wie folgt vorgegangen werden
from sklearn import datasets # Ermöglicht Zugriff auf Sample Dataset
from sklearn import svm # Läd den Estimator Support vector classifcation
iris = datasets.load_iris() # Iris ist ein Sample Dataset von Scikit
digits = datasets.load_digits() # Ist ebenfalls ein Sample Dataset

clf = svm.SVC(gamma=0.001, C=100.)

#### Erstellung eines Train- & Testdatensets####

**Methode mit Zufallsgenerator**

In [0]:
'''Scikit-Learn eignet sich sehr gut für die Splitung eines Train- & Testdatensets
Vorallem überzeugt die sehr einfach Methode der Zufallsgereration welche
im Gegensatz zu anderen Methoden immer die gleichen Sets generiert.''' 

import pandas as pd
from sklearn.model_selection import train_test_split

housing_data = pd.read_csv('https://raw.githubusercontent.com/ageron/handson-ml/master/datasets/housing/housing.csv')

'''In dieser Funktion erstellen wir ein Trainings- und Testset aus den
aus den geladenen Daten. 

Der Wert test_size=0.2 representiert 20%, bedeutet also das 20% der bestehenden 
Daten als Testset verwendet wird. Der Wert random_state definiert den Seed
und garantiert, dass immer das gleiche Testset generiert wird. Wird der Wert
geändert dann verändert sich auch das Testset.'''

train_set, test_set = train_test_split(housing_data, test_size=0.2, random_state=42)

print(len(train_set),len(test_set))

train_set.head()

# Der Vorteil dieser Methode ist, das ein gleichbleibende Zufallauswahl von 
# Testdaten generiert wird. Der Nachteil ist, dass wir nicht wissen wie aus 
# Sicht der Merkmale das Datenset aufgeteilt wurden. Da es bei einer einseitigen
# Verteilung eines kleinen Datensets wie diesem zu erhebelichem Stichproben-
# Bias kommen kann


**Methode mit merkmalbasierter Auswahl**

In [0]:
'''Im folgenden Beispiel generieren wir Train- & Testdatenset aus Merkmalen
um einen Stichproben-Bias zu vermeiden.'''

import pandas as pd
import numpy as np
from sklearn.model_selection import StratifiedShuffleSplit

housing_data = pd.read_csv('https://raw.githubusercontent.com/ageron/handson-ml/master/datasets/housing/housing.csv')

'''Zuerst erstellen wir eine kategorisches Merkmal für das Einkommen indem
wir das mittlere Einkommen (median_income) durch 1.5 teilen damit wir die Anzahl
der Einkommenskategorien begrenzt'''

housing_data["income_cat"] = np.ceil(housing_data["median_income"] / 1.5)
housing_data["income_cat"].where(housing_data["income_cat"] < 5, 5.0, inplace=True)


split = StratifiedShuffleSplit(n_splits=1, test_size=0.2, random_state=42)
for train_index, test_index in split.split(housing_data, housing_data["income_cat"]):
  strat_train_set = housing_data.loc[train_index]
  strat_test_set = housing_data.loc[test_index]

housing_data["income_cat"].value_counts() / len(housing_data)



##Feature Engineering##

Ein wichtiger Aspekt der Datenaufbereitung ist das Verstehen der Daten uns deren Beziehungen. 
Nachfolgenden ein paar unerlässliche Operationen in Bezug auf die Datenaufbereitung. Wir werden hierzu Pandas verwenden.

###Extract, Transfer and Load (ETL)###

Im nachfolgenden Beispiel werden werden wir folgende Schritte ausführen:

1.   Verschiedene Datensätze aus CSV's laden
2.   Die Datensätze vergleichen
3.   Features aus verschiedenen Datensätzen zusammenführen
4.   Die Zusammengeführten Datensätze wieder in ein CSV exportieren.



In [0]:
#Zwei Datensätze laden#

import pandas as pd
import numpy as np
import datetime

daturl = 'https://raw.githubusercontent.com/sakuronohana/my_datascience/master/plane_crash_data.csv'

dimport1 = pd.DataFrame(pd.read_csv(daturl))
dimport2 = pd.DataFrame(pd.read_csv(daturl))


In [0]:
# Datensätze vergleichen #

# Datentypen und Features prüfen
# dimport1.info()
# dimport2.info()

# Datentypen wie Datum und Zeit anpassen
dimport1['date'] = pd.to_datetime(dimport1['date']).dt.strftime('%d.%m.%Y')

# Bestimmte Werte ersetzen
dimport1 = dimport1.replace(to_replace='?', value ='')

# Bestimmter Wert in Spalte suchen
dimport1[dimport1['time'].str.contains('c', na=False)]# mit na=False werden NAN Werte ignoriert
#dimport1[dimport1['time'].str.replace('c','')]

         
#dimport1['time'] = pd.Timestamp(dimport1['time']).time('%m:%h')

## Python mit GUI ##

Auch mit Python besteht die Möglichkeit eine Benutzeroberfläche zu gestallten. Ob wohl dies bei der normalen Data Science Arbeit nicht zwingend notwendig ist, ist das Wissen hier zu dennoch nützlich.

Nachfolgend wird die einer oder andere Library vorgestellt. 

#### Ipywidgets ####

Ist eine Library für Python mit welcher man Widgets [witschet] (Button, Dropdowns usw.) im Jupyter Notebook erstellen kann. Hier eine detailierte Anleitung [Link](https://towardsdatascience.com/bring-your-jupyter-notebook-to-life-with-interactive-widgets-bc12e03f0916). 

Eine Übersicht aller Widgets findet man unter diesem [Link](https://ipywidgets.readthedocs.io/en/stable/examples/Widget%20List.html) 

In [0]:
# Installation von Bamboolib 

!pip install ipywidgets

In [0]:
# Damit es im Jupyter Notebook funktioniert müssen noch verschiedene Extensions geladen werden.

jupyter nbextension enable --py widgetsnbextension

In [0]:
# Import Ipywidgets

import ipywidgets as widgets
from IPython.display import display

In [0]:
# So erstellt man einen Slider

widgets.IntSlider(
    min=0,
    max=10,
    step=1,
    description='Slider:',
    value=3
)

IntSlider(value=3, description='Slider:', max=10)

In [0]:
# Und so gibt man die Zahlen des Slider aus



slider = widgets.IntSlider()
text = widgets.IntText()

display(slider, text)

widgets.jslink((slider, 'value'),(text, 'value'))


IntSlider(value=0)

IntText(value=0)

Link(source=(IntSlider(value=0), 'value'), target=(IntText(value=0), 'value'))

In [0]:
print(dir(widgets))

['Accordion', 'AppLayout', 'Audio', 'BoundedFloatText', 'BoundedIntText', 'Box', 'Button', 'ButtonStyle', 'CallbackDispatcher', 'Checkbox', 'Color', 'ColorPicker', 'Combobox', 'Controller', 'CoreWidget', 'DOMWidget', 'DatePicker', 'Datetime', 'Dropdown', 'FileUpload', 'FloatLogSlider', 'FloatProgress', 'FloatRangeSlider', 'FloatSlider', 'FloatText', 'GridBox', 'GridspecLayout', 'HBox', 'HTML', 'HTMLMath', 'Image', 'IntProgress', 'IntRangeSlider', 'IntSlider', 'IntText', 'Label', 'Layout', 'NumberFormat', 'Output', 'Password', 'Play', 'RadioButtons', 'Select', 'SelectMultiple', 'SelectionRangeSlider', 'SelectionSlider', 'SliderStyle', 'Style', 'Tab', 'Text', 'Textarea', 'ToggleButton', 'ToggleButtons', 'ToggleButtonsStyle', 'TwoByTwoLayout', 'VBox', 'Valid', 'ValueWidget', 'Video', 'Widget', '__builtins__', '__cached__', '__doc__', '__file__', '__jupyter_widgets_base_version__', '__jupyter_widgets_controls_version__', '__loader__', '__name__', '__package__', '__path__', '__protocol_vers

In [0]:
# Erstellen eines Buttons und einer Click-Aktion

btn = widgets.Button(description='Medium')
display(btn)
def btn_eventhandler(obj):
    print('Hello from the {} button!'.format(obj.description))
btn.on_click(btn_eventhandler)

Button(description='Medium', style=ButtonStyle())

Hello from the Medium button!


In [0]:
# Zusammenarbeit Widgets und Dataframes

import pandas as pd
import numpy as np
url = "https://data.london.gov.uk/download/number-international-visitors-london/b1e0f953-4c8a-4b45-95f5-e0d143d5641e/international-visitors-london-raw.csv"
df_london = pd.read_csv(url)

ALL = 'ALL'
def unique_sorted_values_plus_ALL(array):
    unique = array.unique().tolist()
    unique.sort()
    unique.insert(0, ALL)
    return unique

dropdown_year = widgets.Dropdown(options = unique_sorted_values_plus_ALL(df_london.year))

def dropdown_year_eventhandler(change):
    if (change.new == ALL):
        display(df_london)
    else:
        display(df_london[df_london.year == change.new])

dropdown_year.observe(dropdown_year_eventhandler, names='value')

display(dropdown_year)

Dropdown(options=('ALL', 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2…

year quarter        market  ...     spend      nights sample
6637   2004      Q1       Belgium  ...  1.824694    7.118778      5
6638   2004      Q1       Belgium  ...  2.124391   11.397926     13
6639   2004      Q1       Belgium  ...  0.040518    1.053520      1
6640   2004      Q1       Belgium  ...  0.118981    0.760264      1
6641   2004      Q1       Belgium  ...  1.340740   29.719000      6
...     ...     ...           ...  ...       ...         ...    ...
10153  2004      Q4  Other Africa  ...  0.429408   11.692750      2
10154  2004      Q4  Other Africa  ...  2.196219   29.084010      3
10155  2004      Q4  Other Africa  ...  2.211225  221.007949     11
10156  2004      Q4  Other Africa  ...  4.917840  140.509995      1
10157  2004      Q4  Other Africa  ...  0.315062   48.518200      1

[3521 rows x 11 columns]

year quarter        market  ...     spend      nights sample
23507  2009      Q1       Belgium  ...  1.658135    3.526905      4
23508  2009      Q1       Belgium  ...  0.005802    0.193414      1
23509  2009      Q1       Belgium  ...  1.114405   10.390720      4
23510  2009      Q1       Belgium  ...  0.250991    4.365210      1
23511  2009      Q1       Belgium  ...  6.086418   28.973310     18
...     ...     ...           ...  ...       ...         ...    ...
26913  2009      Q4  Other Africa  ...  1.275518   14.345180      4
26914  2009      Q4  Other Africa  ...  1.286421   23.571100      2
26915  2009      Q4  Other Africa  ...  2.121020    7.336200      1
26916  2009      Q4  Other Africa  ...  1.396875  124.607882      6
26917  2009      Q4  Other Africa  ...  4.842701   11.574530      2

[3411 rows x 11 columns]